File: backend/src/main/java/com/example/backend/entity/Recipe.java
package com.example.backend.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.Map;
import java.util.Objects;

@Getter
@Setter
@NoArgsConstructor
public class Recipe {

    public Recipe(InternalRecipeDTO internalRecipeDTO) {
        external = false;
        author = internalRecipeDTO.getAuthor();
        dateAdded = internalRecipeDTO.getDateAdded();
        id = internalRecipeDTO.getId();
        readyInMinutes = internalRecipeDTO.getReadyInMinutes();
        steps = internalRecipeDTO.getSteps();
        calories = ((Number) internalRecipeDTO.getCalories()).doubleValue();
        proteins = ((Number) internalRecipeDTO.getProteins()).doubleValue();
        carbohydrates = ((Number) internalRecipeDTO.getCarbohydrates()).doubleValue();
        fats = ((Number) internalRecipeDTO.getFats()).doubleValue();

        Map<String, Map<String, Object>> map = internalRecipeDTO.getIngredients_map();

        for (Map.Entry<String, Map<String, Object>> entry : map.entrySet()) {
            Ingredient ingredient = new Ingredient();
            ingredient.setName(entry.getKey());
            ingredient.setId(((Number) entry.getValue().get("id")).intValue());
            ingredient.setAmount(((Number) entry.getValue().get("amount")).doubleValue());
            ingredient.setUnit((String) entry.getValue().get("unit"));
            ingredients.add(ingredient);
        }
    }

    public Recipe(ExternalRecipeDTO externalRecipeDTO) {
        external = true;
        author = "";
        dateAdded = "";
        id = externalRecipeDTO.getId();
        title = externalRecipeDTO.getTitle();
        readyInMinutes = externalRecipeDTO.getReadyInMinutes();
        steps = externalRecipeDTO.getSteps();
        calories = externalRecipeDTO.getCalories();
        proteins = externalRecipeDTO.getProteins();
        carbohydrates = externalRecipeDTO.getCarbohydrates();
        fats = externalRecipeDTO.getFats();
        ingredients = externalRecipeDTO.getIngredients();
        imageURL = externalRecipeDTO.getImageURL();
        servings = externalRecipeDTO.getServings();
    }
    private int id;
    private boolean external;
    private String title;

    private ArrayList<Ingredient> ingredients = new ArrayList<>();
    private ArrayList<String> steps;

    private Double readyInMinutes;

    private String author;
    private String dateAdded;
    private String imageURL;
    private int servings;

    // Review information

    private ArrayList<Review> reviews = new ArrayList<>();

    private int amountOfReviews = 0;
    private Double averageReviewScore = (double) 0;

    private Double calories;
    private Double proteins;
    private Double carbohydrates;
    private Double fats;

}

File: frontend/lib/recipe/models/recipe.dart
import 'package:frontend/common/models/ingredient.dart';
import 'package:frontend/recipe/models/review.dart';

class Recipe {
  Recipe(
      {required this.id,
      required this.isExternal,
      required this.title,
      required this.ingredients,
      required this.steps,
      required this.readyInMinutes,
      required this.servings,
      required this.author,
      required this.dateAdded,
      required this.imageURL,
      required this.calories,
      required this.proteins,
      required this.carbohydrates,
      required this.fats,
      required this.reviews,
      required this.amountOfReviews,
      required this.averageReviewScore});

  final int id;
  final bool isExternal;
  final String title;
  final List<Ingredient> ingredients;
  final List<String> steps;
  final double readyInMinutes;
  final String author;
  final String dateAdded;
  final String imageURL;
  final int servings;

  final double calories;
  final double proteins;
  final double carbohydrates;
  final double fats;

  final List<Review> reviews;
  final int amountOfReviews;
  final double averageReviewScore;

  factory Recipe.fromJson(Map<String, dynamic> json) {
    List<String> stepsTemp = (json['steps'] == null)
        ? []
        : (json['steps'] as List).map((i) => i.toString()).toList();
    return Recipe(
      id: json['id'],
      isExternal: json['external'],
      title: json['title'],
      ingredients: (json['ingredients'] as List)
          .map((i) => Ingredient.fromJson(i))
          .toList(), // here
      steps: stepsTemp, // here
      readyInMinutes: json['readyInMinutes'],
      servings: json['servings'],
      author: json['author'] ?? "",
      dateAdded: json['dateAdded'] ?? "",
      imageURL: json['imageURL'],
      calories: json['calories'],
      proteins: json['proteins'],
      carbohydrates: json['carbohydrates'],
      fats: json['fats'],
      //reviews: [],
      reviews: (json['reviews'] as List)
          .map((i) => Review.fromJson(i))
          .toList(), // here
      amountOfReviews: json['amountOfReviews'],
      averageReviewScore: json['averageReviewScore'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'isExternal': isExternal,
    'title': title,
    'author': author,
    'dateAdded': dateAdded,
    'readyInMinutes': readyInMinutes,
    'imageURL': imageURL,
    'servings': servings,
    'calories': calories,
    'proteins': proteins,
    'carbohydrates': carbohydrates,
    'fats': fats,
    'steps': steps,
    'ingredients': ingredients.map((i) => i.toJson()).toList(),
  };
}
